src/components/implementation/no_interface/cFE_booter/ostask.c

```diff
@@ -0,0 +1,815 @@
+		}
+	}
+
+	return OS_ERR_NAME_NOT_FOUND;
+}
+
+int32
+OS_TaskGetInfo(uint32 task_id, OS_task_prop_t *task_prop)
+{
+	if (!task_prop) { return OS_INVALID_POINTER; }
+
+	struct sl_thd *thd = sl_thd_lkup(task_id);
+
+	// TODO: Fix this ugly workaround
+	if (!thd || thd->state == SL_THD_FREE) { return OS_ERR_INVALID_ID; }
+
+	struct cfe_task_info *task_info = &cfe_tasks[task_id];
+	*task_prop = task_info->osal_task_prop;
+
+	return OS_SUCCESS;
+}
+
+/*
+** Main thread waiting API
+*/
+
+/*
+** OS-specific background thread implementation - waits forever for events to occur.
+**
+** This should be called from the BSP main routine / initial thread after all other
+** board / application initialization has taken place and all other tasks are running.
+*/
+void
+OS_IdleLoop(void)
+{
+	while (1) sl_thd_block(0);
+}
+
+/*
+** OS_ApplicationShutdown() provides a means for a user-created thread to request the orderly
+** shutdown of the whole system, such as part of a user-commanded reset command.
+** This is preferred over e.g. ApplicationExit() which exits immediately and does not
+** provide for any means to clean up first.
+*/
+void
+OS_ApplicationShutdown(uint8 flag)
+{
+	PANIC("Unimplemented method!"); // TODO: Implement me!
+}
+
+/*
+** Mutex API
+*/
+
+struct mutex {
+	int               used;
+	struct sl_lock    lock;
+	OS_mut_sem_prop_t prop;
+};
+
+struct sl_lock mutex_data_lock = SL_LOCK_STATIC_INIT();
+struct mutex   mutexes[OS_MAX_MUTEXES];
+
+
+int32
+OS_MutSemCreate(uint32 *sem_id, const char *sem_name, uint32 options)
+{
+	int32 result = OS_SUCCESS;
+
+	sl_lock_take(&mutex_data_lock);
+
+	if (sem_id == NULL || sem_name == NULL) {
+		result = OS_INVALID_POINTER;
+		goto exit;
+	}
+
+	if (!is_valid_name(sem_name)) {
+		result = OS_ERR_NAME_TOO_LONG;
+		goto exit;
+	}
+
+	uint32 id;
+	for (id = 0; id < OS_MAX_MUTEXES; id++) {
+		if (mutexes[id].used && strcmp(sem_name, mutexes[id].prop.name) == 0) {
+			result = OS_ERR_NAME_TAKEN;
+			goto exit;
+		}
+	}
+
+	for (id = 0; id < OS_MAX_MUTEXES; id++) {
+		if (!mutexes[id].used) { break; }
+	}
+	if (id >= OS_MAX_MUTEXES || mutexes[id].used) {
+		result = OS_ERR_NO_FREE_IDS;
+		goto exit;
+	}
+
+	*sem_id = id;
+
+	mutexes[id].used = TRUE;
+	sl_lock_init(&mutexes[id].lock);
+	mutexes[id].prop.creator = sl_thdid();
+	strcpy(mutexes[id].prop.name, sem_name);
+
+exit:
+	sl_lock_release(&mutex_data_lock);
+	return result;
+}
+
+int32
+OS_MutSemGive(uint32 sem_id)
+{
+	int32 result = OS_SUCCESS;
+
+	sl_lock_take(&mutex_data_lock);
+
+	if (sem_id >= OS_MAX_MUTEXES || !mutexes[sem_id].used) {
+		result = OS_ERR_INVALID_ID;
+		goto exit;
+	}
+
+	sl_lock_release(&mutexes[sem_id].lock);
+
+exit:
+	sl_lock_release(&mutex_data_lock);
+
+	return result;
+}
+
+int32
+OS_MutSemTake(uint32 sem_id)
+{
+	int32 result = OS_SUCCESS;
+
+	sl_lock_take(&mutex_data_lock);
+
+	if (sem_id >= OS_MAX_MUTEXES || !mutexes[sem_id].used) {
+		result = OS_ERR_INVALID_ID;
+		goto exit;
+	}
+
+	sl_lock_take(&mutexes[sem_id].lock);
```

Comment:
A simple solution that doesn't really scale in complexity is to just get rid of the `mutex_data_lock`, and take each mutex's lock separately.  Since we're in a static data-structure, this is likely fine.